runtime-sdk/src/enclave_rpc.rs

```diff
@@ -0,0 +1,180 @@
+    core::{
+        consensus::{
+            roothash::Header,
+            state::{
+                beacon::ImmutableState as BeaconState, registry::ImmutableState as RegistryState,
+                roothash::ImmutableState as RoothashState,
+            },
+            verifier::Verifier,
+        },
+        enclave_rpc::{
+            dispatcher::{
+                Dispatcher as RpcDispatcher, Method as RpcMethod,
+                MethodDescriptor as RpcMethodDescriptor,
+            },
+            types::Kind as RpcKind,
+            Context as RpcContext,
+        },
+        future::block_on,
+        protocol::{HostInfo, Protocol},
+        storage::mkvs,
+    },
+    dispatcher,
+    keymanager::KeyManagerClient,
+    module::MethodHandler,
+    state::{self, CurrentState},
+    storage::HostStore,
+    Runtime,
+};
+
+/// Name of the `query` method.
+pub const METHOD_QUERY: &str = "runtime-sdk/query";
+
+/// Arguments for the `query` method.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct QueryRequest {
+    pub round: u64,
+    pub method: String,
+    pub args: Vec<u8>,
+}
+
+/// EnclaveRPC dispatcher wrapper.
+pub(crate) struct Wrapper<R: Runtime> {
+    host_info: HostInfo,
+    host: Arc<Protocol>,
+    key_manager: Option<Arc<KeyManagerClient>>,
+    consensus_verifier: Arc<dyn Verifier>,
+    _runtime: PhantomData<R>,
+}
+
+impl<R> Wrapper<R>
+where
+    R: Runtime + Send + Sync + 'static,
+{
+    pub(crate) fn wrap(
+        rpc: &mut RpcDispatcher,
+        host: Arc<Protocol>,
+        host_info: HostInfo,
+        key_manager: Option<Arc<KeyManagerClient>>,
+        consensus_verifier: Arc<dyn Verifier>,
+    ) {
+        let wrapper = Box::leak(Box::new(Self {
+            host_info,
+            host,
+            key_manager,
+            consensus_verifier,
+            _runtime: PhantomData,
+        }));
+        rpc.add_methods(wrapper.methods());
+    }
+
+    fn methods(&'static self) -> Vec<RpcMethod> {
+        vec![RpcMethod::new(
+            RpcMethodDescriptor {
+                name: METHOD_QUERY.to_string(),
+                kind: RpcKind::NoiseSession,
+            },
+            move |ctx: &_, req: &_| self.rpc_query(ctx, req),
+        )]
+    }
+
+    fn ensure_session_endorsed(&self, ctx: &RpcContext) -> Result<()> {
+        let endorsed_by = ctx
+            .session_info
+            .as_ref()
+            .ok_or(anyhow!("not authorized"))?
+            .endorsed_by
+            .ok_or(anyhow!("not endorsed by host"))?;
+        let host_identity = self
+            .host
+            .get_identity()
+            .ok_or(anyhow!("local identity not available"))?
+            .node_identity()
+            .ok_or(anyhow!("node identity not available"))?;
+        if endorsed_by != host_identity {
+            bail!("not endorsed by host");
+        }
+        Ok(())
+    }
+
+    fn rpc_query(&self, ctx: &RpcContext, req: &QueryRequest) -> Result<Vec<u8>> {
```

Comment:
Do we have test for this? I think that oracle example doesn't test this.

Reply:
It does at least for gas estimation.

Reply:
You are right, should be enough.